Teach the Swift front-end to generate code with sanitizer coverage

delcypher opened the pull request:
<!-- Please complete this template before creating pull request. -->
#### What's in this pull request?

Teach the Swift front-end to generate code with "Sanitizer Coverage" with a new command line flag
``-sanitize-coverage=``. This flag is analogous to Clang's ``-fsanitize-coverage=``.

This instrumentation currently requires ASan or TSan to be enabled
because the module pass created by ``createSanitizerCoverageModulePass()``
inserts calls into functions found in compiler-rt's "sanitizer_common".
"sanitizer_common" is not shipped as an individual library but instead
exists in several of the sanitizer runtime libraries so we have to
link with one of them to avoid linking errors.

The rationale between adding this feature is to allow experimentation
with libFuzzer which currently relies on "Sanitizer Coverage"
instrumentation.

CC: @AnnaZaks @devincoughlin

* * * *

<!-- This selection should only be completed by Swift admin -->
Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please smoke test
        OS X platform               | @swift-ci Please smoke test OS X platform
        Linux platform              | @swift-ci Please smoke test Linux platform

 **Validation Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please test
        OS X platform               | @swift-ci Please test OS X platform
        Linux platform              | @swift-ci Please test Linux platform

Note: Only members of the Apple organization can trigger swift-ci.
</details>
<!-- Thank you for your contribution to Swift! -->

delcypher commented:
@jrose-apple I suspect you will have opinions on this PR. Please take a look when you have time.

jrose-apple commented on lib/IRGen/IRGen.cpp:
```diff
@@ -101,6 +101,15 @@ static void addThreadSanitizerPass(const PassManagerBuilder &Builder,
+                                     legacy::PassManagerBase &PM) {
+  SanitizerCoverageOptions opts;
+  opts.CoverageType = sanitizerCoverageType;
+  assert(opts.CoverageType != SanitizerCoverageOptions::SCK_None);
```
You should be able to make this a `static_assert`, since it's a template parameter.

jrose-apple commented on lib/IRGen/IRGen.cpp:
```diff
@@ -169,6 +178,29 @@ void swift::performLLVMOptimizations(IRGenOptions &Opts, llvm::Module *Module,
+      extensionFn =
+          addSanitizerCoveragePass<SanitizerCoverageOptions::SCK_Edge>;
+      break;
+    default:
```
Rather than a default case, please just add a case for `None`. That way we'll get a warning when someone adds a new coverage kind.

delcypher replied:
The switch is guarded by ``if (Opts.SanitizeCoverage != SanitizerCoverageKind::None)`` so adding a ``case SanitizerCoverageKind::None`` would not be useful because the case would not be reachable.

I could drop the guard add ``case SanitizerCoverageKind::None`` to do nothing in the switch statement but then I would need to guard the ``PMBuilder.addExtension(..)`` calls.

jrose-apple replied:
It's fine to mark the `None` case as `llvm_unreachable("checked above")` or similar. The point is to avoid the `default`.

jrose-apple commented on lib/Option/SanitizerOptions.cpp:
```diff
@@ -34,6 +34,44 @@ static StringRef toStringRef(const SanitizerKind kind) {
+    optionUsed.append(A->getValue(0));
+    Diags.diagnose(SourceLoc(), diag::error_argument_not_allowed_with,
+                   optionUsed,
+                   "Sanitizer disabled. Please use with -sanitize=");
```
I think we should just make a new diagnostic for this, rather than trying to cram the information into an existing diagnostic.

jrose-apple commented on lib/Option/SanitizerOptions.cpp:
```diff
@@ -34,6 +34,44 @@ static StringRef toStringRef(const SanitizerKind kind) {
+  if (kind != SanitizerCoverageKind::None && sanitizer == SanitizerKind::None) {
+    SmallString<128> optionUsed;
+    optionUsed.append(A->getOption().getPrefixedName());
+    optionUsed.append(A->getValue(0));
```
`A->getSpelling` is probably good enough for this, rather than building a new string.

jrose-apple commented on test/Sanitizers/sanitizer_coverage.swift:
```diff
@@ -0,0 +1,17 @@
+// RUN: mkdir %t_coverage_dir
+// RUN: ASAN_OPTIONS=coverage=1,coverage_dir=%t_coverage_dir %t_binary
+// check the coverage file exists
+// RUN: cat %t_coverage_dir/sanitizer_coverage.swift.tmp_binary.*.sancov > /dev/null
```
If you're just checking existence, `ls` is probably better.

jrose-apple commented on test/Sanitizers/sanitizer_coverage.swift:
```diff
@@ -0,0 +1,17 @@
+// RUN: cat %t_coverage_dir/sanitizer_coverage.swift.tmp_binary.*.sancov > /dev/null
+
+// REQUIRES: executable_test
+// For now restrict this test to platforms where I know this test will pass
```
Nitpick: please try to avoid "I" in files/comments. You could use "we" to mean "the project, collectively", or you could rephrase to avoid mentioning a "knower" at all.

AnnaZaks replied:
By default, not everyone checks out compiler-rt (even though CI does).
So you would need to make this test conditional on the existence of asan. You can do that by adding:
// REQUIRES: asan_runtime

See swift/test/lit.cfg for more options.

delcypher replied:
@AnnaZaks Thanks for the tip. Will do.

jrose-apple commented on lib/Driver/Driver.cpp:
```diff
@@ -1104,6 +1104,11 @@ void Driver::buildOutputInfo(const ToolChain &TC, const DerivedArgList &Args,
   if (const Arg *A = Args.getLastArg(options::OPT_sanitize_EQ))
     OI.SelectedSanitizer = parseSanitizerArgValues(A, TC.getTriple(), Diags);
+
+  // Check that the sanitizer coverage flags are supported if supplied
```
Nitpick: please end comments with a period.

jrose-apple commented on include/swift/Option/Options.td:
```diff
@@ -466,4 +466,8 @@ def sanitize_EQ : CommaJoined<["-"], "sanitize=">,
 
+def sanitize_coverage_EQ : Joined<["-"], "sanitize-coverage=">,
+  Flags<[FrontendOption, NoInteractiveOption]>,
+  MetaVarName<"<coverage_type>">, HelpText<"Specify sanitize coverage type">;
```
Nitpick: I'm not super happy with this help text. At the very least it should be "sanitizer coverage", right? But even then it doesn't explain what the flag is. I'm not 100% sure what it is myself, but maybe something like "Track coverage of sanitized code".

(I'd also change the MetaVarName to just "type", because "coverage" is already in the flag name.)

delcypher replied:
> At the very least it should be "sanitizer coverage", right?
Oops. Yes it should be.

The help text for Clang's ``-fsanitize-coverage=`` is "Specify the type of coverage instrumentation for Sanitizers". Should I just use that?

jrose-apple replied:
Sure, that's a bit better. I guess command-line help text isn't supposed to be the full documentation anyway.

jrose-apple commented on include/swift/Basic/Sanitizers.h:
```diff
@@ -21,6 +21,8 @@ enum class SanitizerKind : unsigned {
   Thread,
 };
 
+enum class SanitizerCoverageKind : unsigned { None = 0, Function, BB, Edge };
```
Nitpick: can you split this onto multiple lines? Putting it all on one line doesn't really buy us anything.

delcypher replied:
@jrose-apple It originally was on multiple lines but clang-format decided it knew better. I could change it but if someone clang formats the file in the future it is going to become a single line declaration again.

devincoughlin replied:
Perhaps a solution to the formatting quandary is to add comments describing the coverage kinds?

delcypher replied:
@devincoughlin That's a good idea. More documentation is usually good :)

jrose-apple commented:
This looks really good, Dan. I peppered it with some minor comments but I don't see any real issues.

delcypher commented:
@jrose-apple Thank you for the review. It turns out I need to extend the ``-sanitize-coverage=`` flag to support a few more options (namely ``trace-pc``, `trace-cmp`` and ``indirect-calls``) to match what clang supports. So I think I'll try to fix your issues first and then adds the extra flags I need. These will be all separate commits. Once you're happy with them I can squash everything into a single commit to be merged.

jrose-apple commented:
Sounds good, thanks!

delcypher commented:
@jrose-apple You can probably take another look now. 

Something I'm not very happy about is the introduction of the ``swift::SanitizerCoverageOpts`` struct which is basically the same as ``llvm::SanitizerCoverageOptions``. It only exists to avoid doing ``#include "llvm/Transforms/Instrumentation.h"`` in ``IRGenOptions.h``. In that header file it looks like there aren't any LLVM includes other than what's in the Support library and I don't want to introduce new includes.

AnnaZaks commented:
> Something I'm not very happy about is the introduction of the swift::SanitizerCoverageOpts struct 

I am leaning toward including the definitions from LLVM rather than copying them over, especially, if what's defined in the LLVM header is identical to what you use. You could also consider splitting the LLVM header into 2. For example, one of them could contain only the definitions of the options.

delcypher commented:
@AnnaZaks 
> I am leaning toward including the definitions from LLVM rather than copying them over, especially, if what's defined in the LLVM header is identical to what you use.

I should probably go for that then unless @jrose-apple has objections.

> You could also consider splitting the LLVM header into 2. For example, one of them could contain only the definitions of the options.
Unfortunately I don't think that is practical. The change would need to be made upstream (I don't want this to be a swift only change to LLVM) which might get some push back and take a while to get accepted upstream which would delay getting this PR merged.

jrose-apple commented on include/swift/AST/DiagnosticsFrontend.def:
```diff
@@ -58,6 +58,14 @@ ERROR(error_unsupported_opt_for_target, none,
+
+ERROR(error_option_missing_required_argument, none,
+  "option '%0' is missing a required argument. The required "
+  "argument %1", (StringRef, StringRef))
```
:-/ I'm still unhappy with one of the diagnostic arguments being a whole phrase. How about `"option '%0' is missing a required argument (%1)"`? That way there's less of a grammatical restriction, and it's also more concise.

delcypher replied:
SGTM. I will make this change.

jrose-apple commented on include/swift/Basic/Sanitizers.h:
```diff
@@ -21,6 +21,26 @@ enum class SanitizerKind : unsigned {
+    Edge,     ///< Record coverage for basic block edges
+  } CoverageType : 2;
+  // Experimental features
+  bool IndirectCalls : 1; ///< Record coverage for indirect caller-callee pairs
```
This might be cargo-culting, but IIRC MSVC has odd behavior with boolean bitfields, and we've avoided them elsewhere in the codebase (using `unsigned` instead).

delcypher replied:
Okay that's good to know. It's a moot point now though given that I'm just going to use the struct in LLVM's headers.

jrose-apple commented on lib/Option/SanitizerOptions.cpp:
```diff
@@ -34,6 +34,66 @@ static StringRef toStringRef(const SanitizerKind kind) {
+
+  if (opts.CoverageType == SanitizerCoverageOpts::None) {
+    Diags.diagnose(SourceLoc(), diag::error_option_missing_required_argument,
+                   A->getSpelling(), "is func, bb or edge");
```
I'd suggest quoting "func", "bb", and "edge" here, too.

jrose-apple commented on lib/IRGen/IRGen.cpp:
```diff
@@ -66,6 +66,14 @@ using namespace swift;
 using namespace llvm;
 
+// We need this to access IRGenOptions from extension functions
+class PassManagerBuilderWrapper : public PassManagerBuilder {
```
Please wrap this in an anonymous namespace.

delcypher replied:
Good call.

jrose-apple commented on lib/IRGen/IRGen.cpp:
```diff
@@ -66,6 +66,14 @@ using namespace swift;
+public:
+  const IRGenOptions &IRGOpts;
+  PassManagerBuilderWrapper(const IRGenOptions &IRGOpts)
+      : PassManagerBuilder(), IRGOpts(IRGOpts){};
```
Stray semicolon here (though harmless).

jrose-apple commented:
I'm torn about the LLVM header. That's not (transitively) a small header, but Anna's right that the duplication isn't worth it. Let's include it for now, but file a bug to split it up upstream as suggested.

delcypher commented:
Okay. I've addressed @jrose-apple comments (apart from the LLVM header issue. I'm waiting on @rjmccall to give me the okay on that). Looks like this PR won't merge cleanly anymore so I'm going to rebase now and squash everything into a single commit.

AnnaZaks commented:
For sanitizer tests that crash, please use the following option. This will ensure they do not crash, which is bad because it will generate a crash report. 
env ASAN_OPTIONS=abort_on_error=0

delcypher commented:
@AnnaZaks 
> For sanitizer tests that crash, please use the following option. This will ensure they do not crash, which is bad because it will generate a crash report. 
env ASAN_OPTIONS=abort_on_error=0

Is that necessary here? The single test that use the sanitizer shouldn't be doing anything bad with addresses. The ``test/Sanitizers/sanitizer_coverage.swift`` is only using ASan because we need bits of the ASan compiler-rt linked in for the Sanitizer coverage to be linked in.

delcypher commented:
@jrose-apple @AnnaZaks This PR has been rebased, modified to use the LLVM header and squashed into a single commit. When you have time please take a look.

AnnaZaks commented:
You don't have to add that if the test is not reporting any issues. Though, we might want to have it for consistency.

jrose-apple commented on include/swift/AST/IRGenOptions.h:
```diff
@@ -91,6 +94,9 @@ class IRGenOptions {
   SanitizerKind Sanitize : 2;
 
+  /// Which sanitizer coverage is turned on.
+  llvm::SanitizerCoverageOptions SanitizeCoverage;
```
Oops, nitpick: please move this out of the bitfields, so it doesn't stop them from getting merged.

jrose-apple commented on include/swift/AST/DiagnosticsFrontend.def:
```diff
@@ -58,6 +58,13 @@ ERROR(error_unsupported_opt_for_target, none,
 
+ERROR(error_option_requires_sanitizer, none,
+  "option '%0' requires a sanitizer to be enabled. Use -sanitize= to enable a"
+  " sanitzer", (StringRef))
```
Typo: "sanitzer"

jrose-apple commented on lib/Option/SanitizerOptions.cpp:
```diff
@@ -34,6 +34,68 @@ static StringRef toStringRef(const SanitizerKind kind) {
+  if (opts.CoverageType == llvm::SanitizerCoverageOptions::SCK_None) {
+    Diags.diagnose(SourceLoc(), diag::error_option_missing_required_argument,
+                   A->getSpelling(),
+                   "The required argument is \"func\", \"bb\" or \"edge\"");
```
Still not happy here; see the comment on the unsquashed commit https://github.com/delcypher/swift/commit/8f18b89e845a21ecdc54a4fe154a9a2c588fd571.

delcypher commented:
@AnnaZaks 
> You don't have to add that if the test is not reporting any issues. Though, we might want to have it for consistency.

Oh looks like you've added a ``test/Sanitizers`` directory since I last looked. I'm glad we picked the same name :). I've set the ``abort_on_error=0`` ``ASAN_OPTIONS`` option although I didn't use ``env``. Is that necessary. It doesn't seem to be on my local machine.

delcypher commented:
@jrose-apple Sorry for the issues. I've tried to fix them in the latest version of the commit.

AnnaZaks commented:
> Oh looks like you've added a test/Sanitizers directory since I last looked. I'm glad we picked the same name :). 

I've checked which directory you use in this patch before adding mine:)

> I've set the abort_on_error=0 ASAN_OPTIONS option although I didn't use env. Is that necessary. It doesn't seem to be on my local machine.

I call %target-run %t_tsan-binary, which does not work without using 'env'.

jrose-apple commented:
Looks good on my end. Thanks for bearing with my feedback!

delcypher commented:
@jrose-apple Great. Is there anything left for me to do (e.g. triggering test runs)?

AnnaZaks commented:
@swift-ci test

delcypher commented:
@AnnaZaks I don't think the [Linux build failure](https://ci.swift.org/job/swift-PR-Linux/1759/consoleFull#1635326738ee1a197b-acac-4b17-83cf-a53b95139a76) is to do with me. It looks like its in lldb.

AnnaZaks commented:
@swift-ci please test and merge

AnnaZaks commented:
@swift-ci please test and merge

jrose-apple commented:
<unknown>:0: error: unsupported option '-sanitize=address' for target 'x86_64-unknown-linux-gnu'

    Failing Tests (2):
        Swift :: Driver/sanitize_coverage.swift
        Swift :: IRGen/sanitize_coverage.swift

That looks like a real issue. Are these tests supposed to have REQUIRES guards?

jrose-apple commented:
`REQUIRES: asan_runtime`, right?

jrose-apple commented on test/IRGen/sanitize_coverage.swift:
```diff
@@ -0,0 +1,30 @@
+// RUN: %target-swift-frontend -emit-ir -sanitize=address -sanitize-coverage=edge,indirect-calls %s | FileCheck %s -check-prefix=SANCOV -check-prefix=SANCOV_INDIRECT_CALLS
+// RUN: %target-swift-frontend -emit-ir -sanitize=address -sanitize-coverage=edge,8bit-counters %s | FileCheck %s -check-prefix=SANCOV -check-prefix=SANCOV_8BIT_COUNTERS
+
+import Foundation
```
Now that I think of it, please import Darwin instead of Foundation, to make it easier to port this test to Linux later.

delcypher replied:
Okay. I'm not sure how that makes porting easier but I'm not familiar enough with these modules to know any better.

jrose-apple replied:
It just means we won't accidentally start introducing dependencies on ObjC and/or Foundation.

delcypher commented:
@jrose-apple 
> That looks like a real issue. Are these tests supposed to have REQUIRES guards?

Yes they should. I completely forgot about this. I can probably use ``REQUIRES: asan_runtime`` although for the driver test this a little misleading because it isn't really that the ``asan_runtime`` is required it is asan support in the frontend. Currently the driver will error out if the host is Linux and ``-sanitize=(address|thread)``. Should I just go ahead and use ``REQUIRES: asan_runtime`` or should I use ``XFAIL: linux``?

AnnaZaks commented:
You need XFAIL: linux. I suspect you do not need REQUIRES: asan_runtime for the driver tests. The failure above will be addressed by the XFAIL. The REQUIRES will make sure that the engineers that do not check out compiler RT will not run into test failures.

delcypher commented:
@AnnaZaks Okay I've made the changes. Could you re-run the tests?

AnnaZaks commented:
@swift-ci please test and merge

delcypher commented:
@jrose-apple @AnnaZaks 
I don't think this build failure is related to this PR.

jrose-apple commented:
Yeah, we're seeing this on master too. I think it's okay to merge.

delcypher commented:
@jrose-apple Okay I'm merging then. Fingers crossed...
